modules/tensor_mechanics/src/userobjects/HEVPFlowRatePowerLawJ2.C

```diff
@@ -0,0 +1,117 @@
+#include "HEVPFlowRatePowerLawJ2.h"
+
+template<>
+InputParameters validParams<HEVPFlowRatePowerLawJ2>()
+{
+  InputParameters params = validParams<HEVPFlowRateUOBase>();
+  params.addParam<Real>("reference_flow_rate",0.001,"Reference flow rate for rate dependent flow");
+  params.addParam<Real>("flow_rate_exponent",10,"Power law exponent in flow rate equation");
+  params.addParam<Real>("flow_rate_tol", 1e3, "Tolerance for flow rate");
+  params.addClassDescription("User object to evaluate power law flow rate and flow direction based on J2");
+
+  return params;
+}
+
+HEVPFlowRatePowerLawJ2::HEVPFlowRatePowerLawJ2(const InputParameters & parameters) :
+    HEVPFlowRateUOBase(parameters),
+    _ref_flow_rate(getParam<Real>("reference_flow_rate")),
+    _flow_rate_exponent(getParam<Real>("flow_rate_exponent")),
+    _flow_rate_tol(getParam<Real>("flow_rate_tol"))
+{
+}
+
+bool
+HEVPFlowRatePowerLawJ2::computeValue(const unsigned int & qp, Real & val) const
+{
+  RankTwoTensor pk2_dev = computePK2Deviatoric(_pk2[qp], _ce[qp]);
+  Real eqv_stress = computeEqvStress(pk2_dev, _ce[qp]);
+  val = std::pow(eqv_stress/_strength[qp], _flow_rate_exponent) * _ref_flow_rate;
+
+  if (val > _flow_rate_tol)
+  {
+#ifdef DEBUG
+    mooseWarning("Flow rate greater than " << _flow_rate_tol << " " << val << " " << eqv_stress << " " << _strength[qp]);
+#endif
+    return false;
+  }
+  return true;
+}
+
+bool
+HEVPFlowRatePowerLawJ2::computeDirection(const unsigned int & qp, RankTwoTensor & val) const
+{
+  RankTwoTensor pk2_dev = computePK2Deviatoric(_pk2[qp], _ce[qp]);
+  Real eqv_stress = computeEqvStress(pk2_dev, _ce[qp]);
+
+  val.zero();
+  if (eqv_stress > 0.0)
+    val = 3.0/(2.0 * eqv_stress) * _ce[qp] * pk2_dev * _ce[qp];
+
+  return true;
+}
+
+bool
+HEVPFlowRatePowerLawJ2::computeDerivative(const unsigned int & qp, const std::string & coupled_var_name, Real & val) const
+{
+  val = 0;
+
+  if (_strength_prop_name == coupled_var_name)
+  {
+    RankTwoTensor pk2_dev = computePK2Deviatoric(_pk2[qp], _ce[qp]);
+    Real eqv_stress = computeEqvStress(pk2_dev, _ce[qp]);
+    val = - _ref_flow_rate * _flow_rate_exponent * std::pow(eqv_stress/_strength[qp],_flow_rate_exponent) * 1/_strength[qp];
```

Comment:
`1/_strength[qp]` --> I would change the numerator to "1.0" just to be consistent with other places where you use floating point constants.